### PR TITLE
Issue #12515: Support the latest pattern matching syntax in switch labels

### PR DESCRIPTION
**Pull request by nrmancuso**
#12515

Looks like we actually lucked out with this change, and can keep backwards compatibility.

It will be easier to do this update before update at https://github.com/checkstyle/checkstyle/pull/12516, the inverse will make testing more difficult.

----------------------------------

### Check Regression Reports

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_checks-only-javadoc-error/index.html

- Symmetrical parsing errors only (difference in line number)

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_part1/index.html

- clean

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_part2/index.html

- clean

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_part3/index.html

- clean

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_part4/index.html

- clean

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_part5/index.html

- clean

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_sevntu-check-regression_part_1/index.html

- clean 

https://nrmancuso.github.io/issue-12512/reports_logs/issue-12515_check_diff_reports_2022_12_22/diff_sevntu-check-regression_part_2/index.html

 - clean

-------------------
### AST Regression Report

https://nrmancuso.github.io/issue-12512/reports_logs/diff/index.html

- Symmetrical parsing errors only (difference in line number)

### CODE REVIEW

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsTrickyWhenUsage.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    }
+
+    static void m2(){
+        int when = Integer.parseInt("when");
+    }
+
+    static String m3(Object x) {
+        return switch (x) {
+            // this is horrible, but it compiles
+            case Integer when when when >= 0 -> when.toString();
```
Comment (nrmancuso) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsTrickyWhenUsage.java:20]: :)

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 2]
```diff
@@ -6215,6 +6215,63 @@ public final class TokenTypes {
...
+     * Java Language Specification, &sect;14.30</a>
+     * @see #LITERAL_SWITCH
+     * @see #PATTERN_VARIABLE_DEF
+     * @see #LITERAL_INSTANCEOF
+     * @see #SWITCH_RULE
+     *
+     * @since 10.6
+     */
+    public static final int LITERAL_WHEN =
+        JavaLanguageLexer.LITERAL_WHEN;
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6273]: @strkkk @pbludov @rnveach @romani let's agree on new token/AST (not much of a choice in this case) before I begin check/ AST regression testing.

File: config/checkstyle_resources_suppressions.xml [hunk 3]
```diff
@@ -175,6 +175,12 @@
   <suppress checks="PackageDeclarationCheck"
             files="[\\/]InputAstTreeStringPrinterTextBlocksEscapesAreOneChar\.java"/>
 
+  <!-- AST regression testing requires complicated input, we should not limit input file length -->
```
Comment (nrmancuso) [inline config/checkstyle_resources_suppressions.xml:178]: I will probably send this as a separate PR to speed things up, just keeping CI green now.

File: config/projects-to-test/openjdk19-excluded.files [hunk 4]
```diff
@@ -40,44 +40,32 @@
...
-  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]SimpleAndGuardPattern.java$"/>
+  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]Domination.java$"/>
 </module>
 <module name="BeforeExecutionExclusionFileFilter">
-  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]Switches.java$"/>
+  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]NoModifiersOnBinding.java$"/>
 </module>
 <module name="BeforeExecutionExclusionFileFilter">
-  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]EnumTypeChanges.java$"/>
+  <property name="fileNamePattern" value="[\\/]test[\\/]langtools[\\/]tools[\\/]javac[\\/]patterns[\\/]PatternCaseErrorRecovery.java$"/>
```
Comment (nrmancuso) [inline config/projects-to-test/openjdk19-excluded.files:67]: My script failed to detect that these were not able to be compiled, so I have moved these suppressions below non-compilable line.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 [hunk 5]
```diff
@@ -205,6 +205,7 @@ LITERAL_YIELD:           'yield';
 LITERAL_NON_SEALED:      'non-sealed';
 LITERAL_SEALED:          'sealed';
 LITERAL_PERMITS:         'permits';
+LITERAL_WHEN:            'when';
```
Comment (pbludov) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:208]: Please add `when` to the pattern for illegal identifiers:
https://github.com/checkstyle/checkstyle/blob/master/config/checkstyle_checks.xml#L789

**Review by pbludov — APPROVED** [review]

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 (hunk 5 above)
Comment (nrmancuso) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:208]: Ticket opened at https://github.com/checkstyle/checkstyle/issues/12553

**Comment by nrmancuso** [comment]
Github, generate site

**Review by strkkk — APPROVED** [review]

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 (hunk 5 above)
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:208]: This is missing inside `tokens {`.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4 [hunk 6]
```diff
@@ -870,9 +870,19 @@ pattern
...
-    : primaryPattern LAND expr
+    : primaryPattern guard expr
     ;
 
+/**
+ * We do not need to enforce what the compiler already does; namely, the '&&' syntax
+ * in case labels was only supported as a preview feature in JDK18 and will fail compilation
+ * now. Guarded patterns in expressions still uses '&&', while case labels now use 'when'.
+ * We can allow both alternatives here, since this will help us to maintain backwards
+ * compatibility and avoid more alternatives/ complexity of maintaining two
```
Comment (rnveach) [inline src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4:881]: `alternatives/ complexity` => `alternatives/complexity`

File: src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java [hunk 7]
```diff
@@ -701,6 +701,9 @@ public void testTokenNumbering() {
         assertWithMessage(message)
              .that(JavaLanguageLexer.PATTERN_DEF)
              .isEqualTo(213);
+        assertWithMessage(message)
+             .that(JavaLanguageLexer.LITERAL_WHEN)
+             .isEqualTo(216);
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java:706]: What happened with 214 - 215?

**Review by rnveach — CHANGES_REQUESTED** [review]

File: src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java (hunk 7 above)
Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java]: Fixed. From not adding the token to the tokens block, it threw the ordering off. Added a test to prevent this in the future.

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4 (hunk 6 above)
Comment (nrmancuso) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageParser.g4]: Done

File: src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4 (hunk 5 above)
Comment (nrmancuso) [reply src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java/JavaLanguageLexer.g4:208]: Done

File: src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java [hunk 8]
```diff
@@ -711,7 +719,48 @@ public void testTokenNumbering() {
...
+        final List<String> allTokenNames = Arrays.stream(nullableSymbolicNames)
+                .filter(Objects::nonNull)
+                .collect(Collectors.toList());
+
+        // Since the following tokens are not declared in the 'tokens' block,
+        // they will always appear last in the list of symbolic names provided
+        // by the vocabulary.
+        final List<String> unusedTokenNames = List.of(
+                // reserved keywords that are not part of the language
+                "LITERAL_CONST", "LITERAL_GOTO",
```
Comment (rnveach) [inline src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java:744]: I'm confused, why are these here if they are not part of the language? Goto is not used?

Comment (nrmancuso) [reply src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java:744]: `goto` and `const` are reserved keywords in the java langauge, but they are not used for historical reasons. I included them as keywords in the lexer for completeness in regards to the language spec, but of course they are never used.

Comment (rnveach) [reply src/test/java/com/puppycrawl/tools/checkstyle/grammar/GeneratedJavaTokenTypesTest.java:744]: Ok, they are not in `TokenTypes`.

**Review by rnveach — APPROVED** [review]

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 9]
```diff
@@ -6215,6 +6215,63 @@ public final class TokenTypes {
...
+     * </pre>
+     *
+     * @see <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30">
+     * Java Language Specification, &sect;14.30</a>
+     * @see #LITERAL_SWITCH
+     * @see #PATTERN_VARIABLE_DEF
+     * @see #LITERAL_INSTANCEOF
+     * @see #SWITCH_RULE
+     *
+     * @since 10.6.1
```
Comment (nrmancuso) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: updated, unless this PR makes 10.7? I am not sure about Checkstyle emotional versioning :)

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: @romani ping. It seems like we need a process when versioning number is changed from the original plan.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: We have only first digit as emotional. All others are strictly following semiver.

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: Ok, then this update should be 10.7 (from https://semver.org/):

> Given a version number MAJOR.MINOR.PATCH, increment the:

>MAJOR version when you make incompatible API changes
MINOR version when you add functionality in a backwards compatible manner
PATCH version when you make backwards compatible bug fixes
Additional labels for pre-release and build metadata are available as extensions to the MAJOR.MINOR.PATCH format.

Comment (romani) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: We need to be more attentive, as after merge of any fix for breaking compatibility, it will be 10.7.0

Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Updated to 10.7.0 since we have some bug fixes

**Comment by nrmancuso** [comment]
@romani ping

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsTrickyWhenUsage.java (hunk 1 above)
Comment (romani) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsTrickyWhenUsage.java:20]: @nrmancuso , you can open bug on jdk.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java [hunk 10]
```diff
@@ -6215,6 +6215,63 @@ public final class TokenTypes {
...
+     * </pre>
+     *
+     * @see <a href="https://docs.oracle.com/javase/specs/jls/se17/html/jls-14.html#jls-14.30">
+     * Java Language Specification, &sect;14.30</a>
+     * @see #LITERAL_SWITCH
+     * @see #PATTERN_VARIABLE_DEF
+     * @see #LITERAL_INSTANCEOF
+     * @see #SWITCH_RULE
+     *
+     * @since 10.7
```
Comment (romani) [inline src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java:6270]: 10.7.0

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsInSwitchLabels.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+                    System.out.println(set);
+            default -> throw new IllegalStateException(
+                    "Unexpected value: " + o);
+        }
+    }
+
+    static String m3(Object o) {
+        // patterns in switch are guarded by 'when'
+        return switch (o) {
+            case List list1 when list1.size() >= 0
```
Comment (romani) [inline src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsInSwitchLabels.java:42]: too much obsession with only `>=` in simple examples.

use some/all others in other examples.

**Review by romani — CHANGES_REQUESTED** [review]
two minors:

File: src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsInSwitchLabels.java (hunk 11 above)
Comment (nrmancuso) [reply src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/grammar/java19/InputJava19PatternsInSwitchLabels.java:42]: Created https://github.com/checkstyle/checkstyle/pull/12531/files#diff-bd5e6e07cacf7956c943d4b551aac1b4875f6697ea16304bd9ef4a2d5fe4d086, all operators in the Java language are covered in most simple switch label possible.

File: src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java (hunk 10 above)
Comment (nrmancuso) [reply src/main/java/com/puppycrawl/tools/checkstyle/api/TokenTypes.java]: Done, issue created at https://github.com/checkstyle/checkstyle/issues/12653 so that we never forget.

**Review by romani — APPROVED** [review]
impressive work.
